Teach run-webkit-tests how to run WPT crash tests

rniwa opened the pull request:
#### b3ec691a75db640d7a32cc4c5505bd5e6d95edf2
<pre>
Teach run-webkit-tests how to run WPT crash tests
<a href="https://bugs.webkit.org/show_bug.cgi?id=242633">https://bugs.webkit.org/show_bug.cgi?id=242633</a>

Reviewed by Youenn Fablet.

This patch adds the support for WPT crash tests whereby which the output of the test is ignored.
Instead, run-webkit-tests will check that the test didn&apos;t result in a crash or a timeout.

* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder_legacy.py:
(LayoutTestFinder.find_tests_by_path): Specify is_wpt_crash_test on a new Test object.
(LayoutTestFinder._is_wpt_test): Extracted.
(LayoutTestFinder._is_wpt_crash_test): Added.
* Tools/Scripts/webkitpy/layout_tests/controllers/single_test_runner.py:
(SingleTestRunner.run):
(SingleTestRunner._run_wpt_crash_test): Added. This method implements the core logic for running
a WPT crash test.
* Tools/Scripts/webkitpy/layout_tests/models/test.py:
(Test): Added a new field, is_wpt_crash_test, to indicate whether this is a WPT crash test or not.
* Tools/Scripts/webkitpy/layout_tests/run_webkit_tests_integrationtest.py:
(RunTest.serial_test_basic):
(RunTest.test_wpt_tests): Added.
(RunTest.test_wpt_tests_rebaseline): Added.
* Tools/Scripts/webkitpy/port/test.py:
(TestInstance.__init__):

Canonical link: <a href="https://commits.webkit.org/252469@main">https://commits.webkit.org/252469@main</a>
</pre>

youennf commented on Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder_legacy.py:
```diff
@@ -112,14 +112,21 @@ def find_tests_by_path(self, paths, device_type=None):
+        if not self._is_wpt_test(test_file):
+            return False
+        filename, _ = self._port._filesystem.splitext(test_file)
+        return filename.endswith('-crash')
```
According https://web-platform-tests.org/writing-tests/crashtest.html, the other possibility for a crash test is to be in a crashtests folder. Can we add it?
For instance LayoutTests/imported/w3c/web-platform-tests/mathml/crashtests

youennf replied:
We could also add https://web-platform-tests.org/writing-tests/crashtest.html link in the change log

rniwa replied:
oh, I missed that.

youennf commented on Tools/Scripts/webkitpy/layout_tests/controllers/single_test_runner.py:
```diff
@@ -147,6 +149,23 @@ def _run_compare_test(self):
+
+        failures = []
+        failures.extend(self._handle_error(driver_output))
+        assert(isinstance(failure, test_failures.FailureCrash) or isinstance(failure, test_failures.FailureTimeout) for failure in failures)
```
We might need a specific handling in WTR to handle async tests like LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-dialog-element/showmodal-shadow-sibling-frame-crash.html.
We need to wait for class=test-wait to be removed.
This could be done as a follow-up though.

rniwa replied:
That most certainly needs to be done in a separate follow up.

rniwa replied:
Doing that in https://github.com/WebKit/WebKit/pull/2405

youennf commented on Tools/Scripts/webkitpy/port/test.py:
```diff
@@ -51,6 +51,7 @@ def __init__(self, name):
         self.error = ''
         self.timeout = False
         self.is_reftest = False
+        self.is_w3c_crash_test = False
```
s/w3c/wpt maybe

youennf commented on Tools/Scripts/webkitpy/layout_tests/run_webkit_tests_integrationtest.py:
```diff
@@ -619,6 +619,78 @@ def test_crash_with_stderr(self):
+                'imported': {
+                    'w3c': {
+                        'web-platform-tests': {
+                            'some': {
```
Let's also add a 'crashtests' folder for testing.

rniwa replied:
Sure.

youennf commented on Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder_legacy.py:
```diff
@@ -112,14 +112,21 @@ def find_tests_by_path(self, paths, device_type=None):
+        if not self._is_wpt_test(test_file):
+            return False
+        filename, _ = self._filesystem.splitext(test_file)
+        return filename.endswith('-crash') or self._filesystem.dirname(test_file).endswith(self._port.TEST_PATH_SEPARATOR + 'crashtests')
```
It is unclear whether /crashtests/subfolder/test.html should also be considered as a crash test.
Worth adding a test whatever we choose.

rniwa replied:
It's not supported for now. I checked WPT but there are no examples of crashtests subdirectories for now. I'll add a test

rniwa replied:
Actually, I'm gonna make it work instead.

rniwa committed:
Teach run-webkit-tests how to run WPT crash tests
https://bugs.webkit.org/show_bug.cgi?id=242633

Reviewed by Youenn Fablet.

This patch adds the support for WPT crash tests whereby which the output of the test is ignored.
Instead, run-webkit-tests will check that the test didn't result in a crash or a timeout.

* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder_legacy.py:
(LayoutTestFinder.find_tests_by_path): Specify is_wpt_crash_test on a new Test object.
(LayoutTestFinder._is_wpt_test): Extracted.
(LayoutTestFinder._is_wpt_crash_test): Added.
* Tools/Scripts/webkitpy/layout_tests/controllers/single_test_runner.py:
(SingleTestRunner.run):
(SingleTestRunner._run_wpt_crash_test): Added. This method implements the core logic for running
a WPT crash test.
* Tools/Scripts/webkitpy/layout_tests/models/test.py:
(Test): Added a new field, is_wpt_crash_test, to indicate whether this is a WPT crash test or not.
* Tools/Scripts/webkitpy/layout_tests/run_webkit_tests_integrationtest.py:
(RunTest.serial_test_basic):
(RunTest.test_wpt_tests): Added.
(RunTest.test_wpt_tests_rebaseline): Added.
* Tools/Scripts/webkitpy/port/test.py:
(TestInstance.__init__):

Canonical link: https://commits.webkit.org/252469@main
